### ISSUE TITLE
[PATCH] make B pseudofork safe

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#14393]
Migrated from [rt.perl.org#123544](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=123544) (status was 'resolved')

Searchable as RT123544$

**Comment by p5pRT** [comment]
### From @bulk88

#### Created by @bulk88



See attached patch\.

<details><summary>Perl Info</summary>

```

Flags:
                    category=core
                    severity=low

Site configuration information for perl 5.21.7:

Configured by Owner at Sat Nov 22 21:54:54 2014.

Summary of my perl5 (revision 5 version 21 subversion 7) configuration:
                  Local Commit: 1bce52df028aabe28c20b2d97949e35c17ea811e
                  Ancestor: 7072da8afeba4c87ae623cd913e274396ffcf1cd
                  Platform:
                    osname=MSWin32, osvers=5.1,
archname=MSWin32-x86-multi-thread
                    uname=''
                    config_args='undef'
                    hint=recommended, useposix=true, d_sigaction=undef
                    useithreads=define, usemultiplicity=define
                    use64bitint=undef, use64bitall=undef,
uselongdouble=undef
                    usemymalloc=n, bincompat5005=undef
                  Compiler:
                    cc='cl', ccflags ='-nologo -GF -W3 -O1 -MD -Zi -DNDEBUG
-G7 -GL
-DWIN32 -D_CONSOLE -DNO_STRICT  -DPERL_TEXTMODE_SCRIPTS
-DPERL_HASH_FUNC_ONE_AT_A_TIME -DNO_MATHOMS -DPERL_IMPLICIT_CONTEXT
-DPERL_IMPLICIT_SYS -DUSE_PERLIO -D_USE_32BIT_TIME_T',
                    optimize='-O1 -MD -Zi -DNDEBUG -G7 -GL',
                    cppflags='-DWIN32'
                    ccversion='13.10.6030', gccversion='', gccosandvers=''
                    intsize=4, longsize=4, ptrsize=4, doublesize=8,
byteorder=1234,
doublekind=3
                    d_longlong=undef, longlongsize=8, d_longdbl=define,
longdblsize=8,
longdblkind=0
                    ivtype='long', ivsize=4, nvtype='double', nvsize=8,
Off_t='__int64',
lseeksize=8
                    alignbytes=8, prototype=define
                  Linker and Libraries:
                    ld='link', ldflags ='-nologo -nodefaultlib -debug
-opt:ref,icf
-ltcg  -libpath:"c:\perl521\lib\CORE"  -machine:x86'
                    libpth="C:\Program Files\Microsoft Visual Studio .NET
2003\VC7\lib"
                    libs=oldnames.lib kernel32.lib user32.lib gdi32.lib
winspool.lib
comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib
netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib  version.lib
odbc32.lib odbccp32.lib comctl32.lib msvcrt.lib
                    perllibs=oldnames.lib kernel32.lib user32.lib gdi32.lib
winspool.lib  comdlg32.lib advapi32.lib shell32.lib ole32.lib
oleaut32.lib  netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib
version.lib odbc32.lib odbccp32.lib comctl32.lib msvcrt.lib
                    libc=msvcrt.lib, so=dll, useshrplib=true,
libperl=perl521.lib
                    gnulibc_version=''
                  Dynamic Linking:
                    dlsrc=dl_win32.xs, dlext=dll, d_dlsymun=undef,
ccdlflags=' '
                    cccdlflags=' ', lddlflags='-dll -nologo -nodefaultlib
-debug
-opt:ref,icf -ltcg  -libpath:"c:\perl521\lib\CORE"  -machine:x86'

Locally applied patches:
                    ce7a4d57d0acca9f39a84d36d708c4505dfe45ca
                    ca0b263f4b167ddf97416f657d79ab5bd3344357
                    08919bf863666074243240abbd19cd1a74cc7b74
                    b8a043377dbf39548709b107a11e5cc2714c0e9a
                    efa855eb5cffb7739616c295dd968d1510efeeb0
                    1d47d0b810e26d9a2f9101fb813bd5b3dd332cc9
                    3faca062ddb056db54f73fa55b0a9d473675dd33
                    0b3e905bda3e75ad948a1213f620656b60807393
                    1b1efc719fde05d215e5a13fb38c03e12a3aab08
                    1bce52df028aabe28c20b2d97949e35c17ea811e


@INC for perl 5.21.7:
                    ..\lib
                    C:/perl521/srcnewb4opt/lib
                    .


Environment for perl 5.21.7:
                    HOME (unset)
                    LANG (unset)
                    LANGUAGE (unset)
                    LD_LIBRARY_PATH (unset)
                    LOGDIR (unset)
                    PATH=C:\WINDOWS\system32;C:\Program Files\Microsoft
Visual
Studio
.NET 2003\Vc7\bin;C:\Program Files\Microsoft Visual Studio .NET
2003\Common7\IDE;C:\WINDOWS;C:\Program Files\Git\cmd;C:\Program
Files\Microsoft Visual Studio .NET 2003\Common7\Tools\bin;C:\perl\bin
                    PERL_BADLANG (unset)
                    PERL_JSON_BACKEND=Cpanel::JSON::XS
                    PERL_YAML_BACKEND=YAML
                    SHELL (unset)





















```
</details>

**Comment by p5pRT** [comment]
### From @bulk88

<details><summary>0001-make-B-psuedofork-safe.patch</summary>

```diff
From 55e35a3b211ce234d1677b8cf5f8b65973db0aca Mon Sep 17 00:00:00 2001
From: Daniel Dragan <bulk88@hotmail.com>
Date: Sun, 4 Jan 2015 22:27:04 -0500
Subject: [PATCH] make B psuedofork safe

Previously B incorrectly used MY_CXT since commit 89ca4ac7af (5.7.2),
there was a MY_CXT declared, but it was never cloned after a win32
psuedofork, negating the whole point of using MY_CXT. This was probably
an oversight since the old code didn't use CLONE method, and would have
been threadsafe only if the module was loaded after a psuedofork/ithread
creation. Rearrange my_cxt_t so there isn't an alignment gap between the 32
bit and 32/64 bit ptrs.

This patch does not address the current lack of CLONE method in
Opcode:: and in File::DosGlob . File::Glob was fixed in commit facf34ef48
DynaLoader in commit 8c472fc1d4 and re:: doesn't use MY_CXT anymore at all.

Failure message of the test before the fix was

not ok 88 - special SV table works after psuedofork
#   Failed test 'special SV table works after psuedofork'
#   at b.t line 229.
#          got: 'B::PVNV'
#     expected: 'B::SPECIAL'
---
 ext/B/B.xs        |   33 ++++++++++++++++++++++++---------
 ext/B/t/b.t       |   13 +++++++++++++
 pod/perldelta.pod |    8 ++++++++
 3 files changed, 45 insertions(+), 9 deletions(-)

diff --git a/ext/B/B.xs b/ext/B/B.xs
index 0f52959..72a33ae 100644
--- a/ext/B/B.xs
+++ b/ext/B/B.xs
@@ -103,8 +103,8 @@ static const size_t opsizes[] = {
 #define MY_CXT_KEY "B::_guts" XS_VERSION
 
 typedef struct {
-    int		x_walkoptree_debug;	/* Flag for walkoptree debug hook */
     SV *	x_specialsv_list[7];
+    int		x_walkoptree_debug;	/* Flag for walkoptree debug hook */
 } my_cxt_t;
 
 START_MY_CXT
@@ -112,6 +112,17 @@ START_MY_CXT
 #define walkoptree_debug	(MY_CXT.x_walkoptree_debug)
 #define specialsv_list		(MY_CXT.x_specialsv_list)
 
+
+static void B_init_my_cxt(pTHX_ my_cxt_t * cxt) {
+    cxt->x_specialsv_list[0] = Nullsv;
+    cxt->x_specialsv_list[1] = &PL_sv_undef;
+    cxt->x_specialsv_list[2] = &PL_sv_yes;
+    cxt->x_specialsv_list[3] = &PL_sv_no;
+    cxt->x_specialsv_list[4] = (SV *) pWARN_ALL;
+    cxt->x_specialsv_list[5] = (SV *) pWARN_NONE;
+    cxt->x_specialsv_list[6] = (SV *) pWARN_STD;
+}
+
 static opclass
 cc_opclass(pTHX_ const OP *o)
 {
@@ -778,14 +789,7 @@ BOOT:
     CV *cv;
     const char *file = __FILE__;
     MY_CXT_INIT;
-    specialsv_list[0] = Nullsv;
-    specialsv_list[1] = &PL_sv_undef;
-    specialsv_list[2] = &PL_sv_yes;
-    specialsv_list[3] = &PL_sv_no;
-    specialsv_list[4] = (SV *) pWARN_ALL;
-    specialsv_list[5] = (SV *) pWARN_NONE;
-    specialsv_list[6] = (SV *) pWARN_STD;
-    
+    B_init_my_cxt(aTHX_ &(MY_CXT));
     cv = newXS("B::init_av", intrpvar_sv_common, file);
     ASSIGN_COMMON_ALIAS(I, initav);
     cv = newXS("B::check_av", intrpvar_sv_common, file);
@@ -975,7 +979,18 @@ threadsv_names()
     PPCODE:
 
 
+#ifdef USE_ITHREADS
+void
+CLONE(...)
+PPCODE:
+    PUTBACK; /* some vars go out of scope now in machine code */
+    {
+	MY_CXT_CLONE;
+	B_init_my_cxt(aTHX_ &(MY_CXT));
+    }
+    return; /* dont execute another implied XSPP PUTBACK */
 
+#endif
 
 MODULE = B	PACKAGE = B::OP
 
diff --git a/ext/B/t/b.t b/ext/B/t/b.t
index 0c1a0f3..06ee025 100644
--- a/ext/B/t/b.t
+++ b/ext/B/t/b.t
@@ -12,6 +12,7 @@ BEGIN {
 $|  = 1;
 use warnings;
 use strict;
+use threads;
 use Test::More;
 
 BEGIN { use_ok( 'B' ); }
@@ -217,6 +218,18 @@ is($gv_ref->GPFLAGS & B::GPf_ALIASED_SV, 0, 'GPFLAGS are unset');
 is(ref B::sv_yes(), "B::SPECIAL", "B::sv_yes()");
 is(ref B::sv_no(), "B::SPECIAL", "B::sv_no()");
 is(ref B::sv_undef(), "B::SPECIAL", "B::sv_undef()");
+SKIP: {
+    skip('no fork', 1)
+	unless ($Config::Config{d_fork} or $Config::Config{d_pseudofork});
+    my $pid;
+    if ($pid = fork) {
+        waitpid($pid,0);
+    }
+    else {
+        is(ref B::svref_2object(\(!!0)), "B::SPECIAL", "special SV table works after psuedofork");
+        exit;
+    }
+}
 
 # More utility functions
 is(B::ppname(0), "pp_null", "Testing ppname (this might break if opnames.h is changed)");
diff --git a/pod/perldelta.pod b/pod/perldelta.pod
index 3a70b71..7ee0ec4 100644
--- a/pod/perldelta.pod
+++ b/pod/perldelta.pod
@@ -147,6 +147,14 @@ XXX
 
 L<XXX> has been upgraded from version A.xx to B.yy.
 
+=item *
+
+L<B> has been upgraded from version 1.54 to 1.55.
+
+A bug where, after an ithread creation or psuedofork, special/immortal SVs in
+the child ithread/psuedoprocess did not have the correct class of
+C<B::SPECIAL>, has been fixed.
+
 =back
 
 =head2 Removed Modules and Pragmata
-- 
1.7.9.msysgit.0


```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Jan 04 19&#8203;:31&#8203;:18 2015\, bulk88 wrote&#8203;:
> This is a bug report for perl from bulk88@&#8203;hotmail\.com\,
> generated with the help of perlbug 1\.40 running under perl 5\.21\.7\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
> 
> See attached patch\.

Thank you\.  I applied this as a462fa007\, but it seems to be causing Jenkins failures\.  See \<1751533573\.300\.1420468728337\.JavaMail\.jenkins@&#8203;jenkins\-001\.ams6\.corp\.booking\.com> for instance\.


\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @bulk88

On Mon Jan 05 08&#8203;:27&#8203;:09 2015\, sprout wrote&#8203;:
> On Sun Jan 04 19&#8203;:31&#8203;:18 2015\, bulk88 wrote&#8203;:
> > This is a bug report for perl from bulk88@&#8203;hotmail\.com\,
> > generated with the help of perlbug 1\.40 running under perl 5\.21\.7\.
> >
> >
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > \[Please describe your issue here\]
> >
> > See attached patch\.
> 
> Thank you\.  I applied this as a462fa007\, but it seems to be causing
> Jenkins failures\.  See
> \<1751533573\.300\.1420468728337\.JavaMail\.jenkins@&#8203;jenkins\-
> 001\.ams6\.corp\.booking\.com> for instance\.

Fix attached\. I am not sure if b\.t will pass \(or fail with "out of sequence"\) on unix since psuedofork is different from fork\. The failure only happened on unthreaded perl\.

\-\- 
bulk88 ~ bulk88 at hotmail\.com

**Comment by p5pRT** [comment]
### From @bulk88

<details><summary>0001-fix-test-fail-on-unthreaded-perl.patch</summary>

```diff
From deee468d15bb874b7f2e597df78229b943e9ccf5 Mon Sep 17 00:00:00 2001
From: Daniel Dragan <bulk88@hotmail.com>
Date: Mon, 5 Jan 2015 14:57:30 -0500
Subject: [PATCH] fix test fail on unthreaded perl

../ext/B/t/b.t                          (Wstat: 65280 Tests: 0 Failed: 0)
  Non-zero exit status: 255
  Parse errors: No plan found in TAP output

part of [perl #123544]
---
 ext/B/t/b.t |    5 ++++-
 1 files changed, 4 insertions(+), 1 deletions(-)

diff --git a/ext/B/t/b.t b/ext/B/t/b.t
index 06ee025..ed60847 100644
--- a/ext/B/t/b.t
+++ b/ext/B/t/b.t
@@ -12,7 +12,10 @@ BEGIN {
 $|  = 1;
 use warnings;
 use strict;
-use threads;
+use Config;
+BEGIN  {
+    eval { require threads; threads->import; }
+}
 use Test::More;
 
 BEGIN { use_ok( 'B' ); }
-- 
1.7.9.msysgit.0


```
</details>

**Comment by p5pRT** [comment]
### From @bulk88

On Mon Jan 05 12&#8203;:01&#8203;:19 2015\, bulk88 wrote&#8203;:
> Fix attached\. I am not sure if b\.t will pass \(or fail with "out of
> sequence"\) on unix since psuedofork is different from fork\. The
> failure only happened on unthreaded perl\.

Revised patch attached\. There was an unneeded "use Config" in the first one\.

\-\- 
bulk88 ~ bulk88 at hotmail\.com

**Comment by p5pRT** [comment]
### From @bulk88

On Mon Jan 05 12&#8203;:03&#8203;:10 2015\, bulk88 wrote&#8203;:
> On Mon Jan 05 12&#8203;:01&#8203;:19 2015\, bulk88 wrote&#8203;:
> > Fix attached\. I am not sure if b\.t will pass \(or fail with "out of
> > sequence"\) on unix since psuedofork is different from fork\. The
> > failure only happened on unthreaded perl\.
> 
> Revised patch attached\. There was an unneeded "use Config" in the first one\.

not attached\, now attached

\-\- 
bulk88 ~ bulk88 at hotmail\.com

**Comment by p5pRT** [comment]
### From @bulk88

<details><summary>0001-fix-test-fail-on-unthreaded-perl.patch</summary>

```diff
From eb2f48333bc894085fdf901de70cd59b0d99e638 Mon Sep 17 00:00:00 2001
From: Daniel Dragan <bulk88@hotmail.com>
Date: Mon, 5 Jan 2015 15:01:04 -0500
Subject: [PATCH] fix test fail on unthreaded perl

../ext/B/t/b.t                          (Wstat: 65280 Tests: 0 Failed: 0)
  Non-zero exit status: 255
  Parse errors: No plan found in TAP output

part of [perl #123544]
---
 ext/B/t/b.t |    4 +++-
 1 files changed, 3 insertions(+), 1 deletions(-)

diff --git a/ext/B/t/b.t b/ext/B/t/b.t
index 06ee025..9c061aa 100644
--- a/ext/B/t/b.t
+++ b/ext/B/t/b.t
@@ -12,7 +12,9 @@ BEGIN {
 $|  = 1;
 use warnings;
 use strict;
-use threads;
+BEGIN  {
+    eval { require threads; threads->import; }
+}
 use Test::More;
 
 BEGIN { use_ok( 'B' ); }
-- 
1.7.9.msysgit.0


```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon Jan 05 12&#8203;:03&#8203;:37 2015\, bulk88 wrote&#8203;:
> On Mon Jan 05 12&#8203;:03&#8203;:10 2015\, bulk88 wrote&#8203;:
> > On Mon Jan 05 12&#8203;:01&#8203;:19 2015\, bulk88 wrote&#8203;:
> > > Fix attached\. I am not sure if b\.t will pass \(or fail with "out of
> > > sequence"\) on unix since psuedofork is different from fork\. The
> > > failure only happened on unthreaded perl\.
> >
> > Revised patch attached\. There was an unneeded "use Config" in the
> > first one\.
> 
> not attached\, now attached

Thank you\.  Applied as 200b5b4\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @bulk88

On Mon Jan 05 13&#8203;:55&#8203;:14 2015\, sprout wrote&#8203;:
> On Mon Jan 05 12&#8203;:03&#8203;:37 2015\, bulk88 wrote&#8203;:
> > On Mon Jan 05 12&#8203;:03&#8203;:10 2015\, bulk88 wrote&#8203;:
> > > On Mon Jan 05 12&#8203;:01&#8203;:19 2015\, bulk88 wrote&#8203;:
> > > > Fix attached\. I am not sure if b\.t will pass \(or fail with "out of
> > > > sequence"\) on unix since psuedofork is different from fork\. The
> > > > failure only happened on unthreaded perl\.
> > >
> > > Revised patch attached\. There was an unneeded "use Config" in the
> > > first one\.
> > 
> > not attached\, now attached
> 
> Thank you\.  Applied as 200b5b4\.
> 

Failed again http&#8203;://perl5\.git\.perl\.org&#8203;:8080/job/perl5/5639/console \, with "out of sequence" like I predicted :\(

I've been trying to avoid having b\.t use test\.pl but it looks like there is no other portable way to run a test in a fork other than to use test\.pl\. Test&#8203;::More does not offer fresh\_perl\_is\(\)\. And is "threads&#8203;::shared" "working" across a Win32 psuedofork\, a bug\, not a feature?

\-\- 
bulk88 ~ bulk88 at hotmail\.com

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon Jan 05 12&#8203;:01&#8203;:19 2015\, bulk88 wrote&#8203;:
> Fix attached\. I am not sure if b\.t will pass \(or fail with "out of
> sequence"\) on unix since psuedofork is different from fork\. The
> failure only happened on unthreaded perl\.

I should have paid more attention to your comment\.  I only ran the test manually and did not notice the tests were out of sequence\.  I have just fixed it in 216b3a07\, using a magical option that Test&#8203;::Stream provides\.  \(I really don’t know how it works\.\)

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @bulk88

On Mon Jan 05 14&#8203;:15&#8203;:13 2015\, bulk88 wrote&#8203;:
> Failed again http&#8203;://perl5\.git\.perl\.org&#8203;:8080/job/perl5/5639/console \,
> with "out of sequence" like I predicted :\(
> 
> I've been trying to avoid having b\.t use test\.pl but it looks like
> there is no other portable way to run a test in a fork other than to
> use test\.pl\. Test&#8203;::More does not offer fresh\_perl\_is\(\)\. And is
> "threads&#8203;::shared" "working" across a Win32 psuedofork\, a bug\, not a
> feature?

Maybe adding a plan\(\) to b\.t will fix the out of sequence problem? b\.t currently has no plan\(\)\, just a done\_testing\(\)\.

\-\- 
bulk88 ~ bulk88 at hotmail\.com
